src/nowsecure-snapshot.ts

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright © 2022 NowSecure Inc.
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+import type { DependencySnapshot, Manifest } from "./types/dependency-snapshot";
+import type { Deputy } from "./types/deputy";
+import * as client from "@actions/http-client";
+import { version } from "./nowsecure-version";
+import { USER_AGENT } from "./nowsecure-client";
+
+/**
+ * Contains environmental information for conversion routines.
+ */
+export interface ActionContext {
+  sha: string;
+  ref: string;
+  job: string;
+  runId: number;
+  repo: {
+    owner: string;
+    repo: string;
+  };
+}
+
+/**
+ * Upload URL for snapshots.
+ */
+const snapshotUrl = (owner: string, repo: string) =>
+  `https://api.github.com/repos/${owner}/${repo}/dependency-graph/snapshots`;
+
+function encodePurl(ecosystem: string, name: string, version?: string): string {
+  let purl = `pkg:${ecosystem}/${name}`;
+  if (version) {
+    purl += `@${version}`;
+  }
+
+  return purl;
+}
+
+/**
+ * Convert a Platform report to the Snapshot format.
+ */
+export function convertToSnapshot(
+  deputy: Deputy,
+  { ref, sha, job, runId }: ActionContext
+): DependencySnapshot {
+  // The file name of the application bundle that was scanned.
+  const name = deputy.package;
+  const manifests = new Map<string, Manifest>();
+
+  for (const component of deputy.components) {
+    // If there is a leading slash, remove.
+    let source = component.source;
+    if (source.length > 0 && source[0] === "/") {
+      source = source.substr(1);
+    }
+
+    if (!manifests.has(source)) {
+      const manifest: Manifest = {
+        name: source,
+        resolved: {},
+      };
+
+      manifests.set(source, manifest);
+    }
+
+    const manifest = manifests.get(source);
```

Comment:
Depends on style preference (since performance doesn't matter here), but I usually use this pattern:

```suggestion
    let manifest = manifests.get(source);
    if (!manifest) {
      manifest = {
        name: source,
        resolved: {},
      };
      manifests.set(source, manifest);
    }

```